control/controller_test.go

```diff
@@ -219,3 +260,55 @@ func TestController_BlockedExecutor(t *testing.T) {
 		t.Fatalf("unexpected error: got=%q want=%q", got, want)
 	}
 }
+
+func TestController_Shutdown(t *testing.T) {
+	executor := mock.NewExecutor()
+	ctrl := New(Config{})
+	ctrl.executor = executor
+
+	// Create a bunch of queries and never call Ready which should leave them in the controller.
+	queries := make([]flux.Query, 0, 10)
+	for i := 0; i < 10; i++ {
+		q, err := ctrl.Query(context.Background(), mockCompiler)
+		if err != nil {
+			t.Errorf("unexpected error: %s", err)
+			continue
+		}
+		queries = append(queries, q)
+	}
+
+	// Run shutdown which should wait until the queries are finished.
+	var wg syncutil.WaitGroup
+	wg.Do(func() error {
+		return ctrl.Shutdown(context.Background())
+	})
+
+	// A new query should be rejected.
+	if _, err := ctrl.Query(context.Background(), mockCompiler); err == nil {
+		t.Error("expected error")
+	}
+
+	// There should be 10 active queries.
+	if want, got := 10, len(ctrl.Queries()); want != got {
+		t.Fatalf("unexpected query count -want/+got\n\t- %d\n\t+ %d", want, got)
+	}
+
+	// Mark each of the queries as done.
+	for _, q := range queries {
+		q := q
```

Comment:
Why is it necessary to shadow here?

Reply:
This is because of how for loops interact with closures.

https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables